### PR TITLE
cl-3d-denoise: enable "sub-group" to improve performance

### PR DESCRIPTION
**Pull request by zongwave**
Use  CL "sub-group" extension  to improve performance.

### CODE REVIEW

File: cl_kernel/kernel_3d_denoise_slm.cl [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    int start_x = mad24(group_id_x, WORKGROUP_WIDTH, -REF_BLOCK_X_OFFSET);
+    int start_y = mad24(group_id_y, WORKGROUP_HEIGHT, -REF_BLOCK_Y_OFFSET);
+
+    for (int i = 0; i < REF_BLOCK_HEIGHT; i++) {
+        for (int j = local_id_x; j < REF_BLOCK_WIDTH; j += WORKGROUP_WIDTH) {
+            int corrd_x = start_x + (j % REF_BLOCK_WIDTH);
+            local_ref_cache[i * REF_BLOCK_WIDTH + j] = read_imagef(input, sampler, (int2)(corrd_x, start_y));
+        }
+        start_y += 1;
+    }
```
Comment (windyuan) [inline cl_kernel/kernel_3d_denoise_slm.cl]: double for-loop slow down performance and cause same slm-cache frequently re-write/re-read very bad. please fix it first.
In this case there are 8x8=64 workitems in a workgroup. and would load 10x16=160 pixels. the best load only need each workitem read (160/64 = )3 times. but this double-for-loop need each workitem load 10/2 x 10 = 50times which slow down the performance heavily.

**Comment by windyuan** [comment]
Just take a full look, kernel_3d_denoise_slm.cl is for pure slm version. if this code checkin, we still need a workable slm version. please fix it or remove it.

**Comment by zongwave** [comment]
fix slm version shared local memory load misusage and register spill issue.

**Comment by zongwave** [comment]
calculate sub-group local id with work group local id

**Comment by windyuan** [comment]
merged. thanks.
commit b43f99c0ca903a62838628dc1c06f3d5fb576746

        --- a/modules/ocl/cl_3d_denoise_handler.cpp
        +++ b/modules/ocl/cl_3d_denoise_handler.cpp
        @@ -30,15 +30,21 @@ namespace XCam {
         #define CL_3D_DENOISE_ENABLE_SUBGROUP 1
         #define CL_3D_DENOISE_IIR_FILTERING   1
        
        +#if CL_3D_DENOISE_ENABLE_SUBGROUP
        +#define KERNEL_3D_DENOISE_NAME "kernel_3d_denoise"
        +#else
        +#define KERNEL_3D_DENOISE_NAME "kernel_3d_denoise_slm"
        +#endif
        +
         CL3DDenoiseImageKernel::CL3DDenoiseImageKernel (SmartPtr<CLContext> &context,
                 const char *name,
        		 
        ......
        ......
        
        -#if CL_3D_DENOISE_ENABLE_SUBGROUP
        +//#if CL_3D_DENOISE_ENABLE_SUBGROUP
             XCAM_CL_KERNEL_FUNC_SOURCE_BEGIN(kernel_3d_denoise)
        +#if CL_3D_DENOISE_ENABLE_SUBGROUP
         #include "kernel_3d_denoise.clx"
        -    XCAM_CL_KERNEL_FUNC_END;
         #else
        -    XCAM_CL_KERNEL_FUNC_SOURCE_BEGIN(kernel_3d_denoise_slm)
         #include "kernel_3d_denoise_slm.clx"
        -    XCAM_CL_KERNEL_FUNC_END;
         #endif
        +    XCAM_CL_KERNEL_FUNC_END;
        +//#else
        +//   XCAM_CL_KERNEL_FUNC_SOURCE_BEGIN(kernel_3d_denoise)
        +//#include "kernel_3d_denoise_slm.clx"
        +//    XCAM_CL_KERNEL_FUNC_END;
        +//#endif
        
             if (channel & CL_IMAGE_CHANNEL_Y) {
        -#if CL_3D_DENOISE_ENABLE_SUBGROUP
        -        denoise_kernel = new CL3DDenoiseImageKernel (context, "kernel_3d_denoise", CL_IMAGE_CHANNEL_Y, denoise_handler);
        +        denoise_kernel = new CL3DDenoiseImageKernel (context, KERNEL_3D_DENOISE_NAME, CL_IMAGE_CHANNEL_Y, denoise_handler);
                 ret = denoise_kernel->load_from_source (
                           kernel_3d_denoise_body, strlen (kernel_3d_denoise_body),
                           NULL, NULL,
                           build_options);
        -#else
        -        denoise_kernel = new CL3DDenoiseImageKernel (context, "kernel_3d_denoise_slm", CL_IMAGE_CHANNEL_Y, denoise_handler);
        -        ret = denoise_kernel->load_from_source (
        -                  kernel_3d_denoise_slm_body, strlen (kernel_3d_denoise_slm_body),
        -                  NULL, NULL,
        -                  build_options);
        -#endif
                 XCAM_FAIL_RETURN (
                     WARNING,
                     ret == XCAM_RETURN_NO_ERROR,
        ......
        ......

**Pull request closed without merging**